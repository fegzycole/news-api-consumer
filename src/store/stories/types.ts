export const SET_STORIES = "SET_STORIES";
export const SET_ERROR = "SET_ERROR";
export const SET_FETCHING_STORIES = "SET_FETCHING_STORIES";

export enum StoryCategories {
  TopHeadlines = "Top Headlines",
  Business = "Business",
  Entertainment = "Entertainment",
  General = "General",
  Health = "Health",
  Science = "Science",
  Sports = "Sports",
  Technology = "Technology",
}

export interface StoryData {
  source: {
    id: null | number;
    name: string;
  };
  author: string;
  title: string;
  description: string;
  url: string;
  urlToImage: string;
  publishedAt: Date | string;
  content: string;
}

export interface StoriesState {
  stories: StoryData[];
  fetchingStories: boolean;
  error?: string;
}

export interface AddStories {
  stories: StoryData[];
  clearStories: boolean;
}

export interface SetStories {
  type: typeof SET_STORIES;
  payload: AddStories;
}

export interface SetError {
  type: typeof SET_ERROR;
  payload: string;
}

export interface SetFetchingStories {
  type: typeof SET_FETCHING_STORIES;
  payload: boolean;
}

export type MoviesActionTypes =
  | SetStories
  | SetError
  | SetFetchingStories;
