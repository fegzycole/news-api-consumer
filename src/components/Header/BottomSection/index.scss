@import "../../../styles/mixins";
@import "../../../styles/palette";

.Categories_Container {
  padding: 1.25rem;
  width: 100%;
  overflow-x: scroll;
}

.Categories {
  width: 800px;
  @include flex-styles(row, space-between);

  @include lgScreen {
    margin: 0 auto;
  }
}

.Category_Btn {
  @include text-styles(500, 14px, $primary-color);
}

.Selected {
  color: $secondary-color;
}
