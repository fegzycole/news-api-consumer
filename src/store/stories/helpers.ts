import { StoryCategories } from "./types";

export const apiKey = process.env.REACT_APP_API_KEY;
export const apiUrl = "https://newsapi.org/v2/top-headlines";

export const getRequestPath = (
  category: StoryCategories,
  pageNumber: number
): string => {
  let path = `${apiUrl}?apiKey=${apiKey}&pageSize=10&page=${pageNumber}&language=en&sortBy=publishedAt`;

  if (category !== StoryCategories.TopHeadlines) {
    path = `${path}&category=${category.toLowerCase()}`;
  }

  return path;
};

export const getResponseError = (error: any): string => {
  let serverErrResponse = error?.response?.data?.message;

  if (!serverErrResponse) {
    serverErrResponse = error.message;
  }

  return serverErrResponse;
};
