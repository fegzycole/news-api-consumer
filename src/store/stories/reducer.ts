import {
  MoviesActionTypes,
  StoriesState,
  SET_STORIES,
  SET_ERROR,
  SET_FETCHING_STORIES,
} from "./types";

const initialState: StoriesState = {
  stories: [],
  fetchingStories: false,
  error: undefined,
};

const storiesReducer = (
  state: StoriesState = initialState,
  action: MoviesActionTypes
) => {
  switch (action.type) {
    case SET_STORIES:
      return {
        ...state,
        stories: action.payload.clearStories
          ? action.payload.stories
          : state.stories.concat(action.payload.stories),
        fetchingStories: false,
      };
    case SET_ERROR:
      return {
        ...state,
        error: action.payload,
        fetchingStories: false,
      };
    case SET_FETCHING_STORIES:
      return {
        ...state,
        fetchingStories: action.payload,
      };
    default:
      return state;
  }
};

export default storiesReducer;
