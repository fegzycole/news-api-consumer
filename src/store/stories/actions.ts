import { Dispatch } from "react";
import axios from "axios";

import {
  StoryData,
  SetStories,
  SET_STORIES,
  SetError,
  SET_ERROR,
  SetFetchingStories,
  SET_FETCHING_STORIES,
  StoryCategories,
  AddStories,
} from "./types";
import { getResponseError, getRequestPath } from "./helpers";

export const setStories = (payload: AddStories): SetStories => {
  return {
    type: SET_STORIES,
    payload,
  };
};

export const setError = (payload: string): SetError => {
  return {
    type: SET_ERROR,
    payload,
  };
};

export const setFetchingStories = (payload: boolean): SetFetchingStories => {
  return {
    type: SET_FETCHING_STORIES,
    payload,
  };
};

export const getStories = (
  category: StoryCategories,
  pageNumber: number
): ((
  dispatch: Dispatch<{ type: string; payload?: unknown }>
) => Promise<void>) => {
  return async (dispatch) => {
    dispatch(setError(""));
    dispatch(setFetchingStories(true));

    const requestPath = getRequestPath(category, pageNumber);

    try {
      const request = await axios.get(requestPath);

      const responseData = request.data.articles as StoryData[];

      const stories: StoryData[] = responseData as StoryData[];

      const payload: AddStories = {
        stories,
        clearStories: pageNumber === 1,
      };

      return dispatch(setStories(payload));
    } catch (error: any) {
      const renderedError = getResponseError(error);
      return dispatch(setError(renderedError));
    }
  };
};
