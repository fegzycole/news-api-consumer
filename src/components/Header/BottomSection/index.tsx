import { v4 as uuidv4 } from "uuid";

import { HeaderProps } from "../index";
import { StoryCategories } from "../../../store/stories/types";
import "./index.scss";

const categories = Object.values(StoryCategories).map((val) => val);

const BottomSection: React.FC<HeaderProps> = ({
  selectedCategory,
  handleClick,
}) => {
  return (
    <nav className="Categories_Container">
      <ul className="Categories">
        {categories.map((category) => (
          <li className="Category" key={uuidv4()}>
            <button
              onClick={() => handleClick(category)}
              className={`Category_Btn ${
                category === selectedCategory && `Selected`
              }`}
            >
              {category}
            </button>
          </li>
        ))}
      </ul>
    </nav>
  );
};

export default BottomSection;
