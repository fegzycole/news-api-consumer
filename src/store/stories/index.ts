import * as Types from './types';
import storiesReducer from './reducer';
import * as Actions from './actions';

export { Types, storiesReducer, Actions };
