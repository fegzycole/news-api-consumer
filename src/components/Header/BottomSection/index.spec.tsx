import { render } from "@testing-library/react";
import { HeaderProps } from "..";
import { StoryCategories } from "../../../store/stories/types";

import BottomSection from "./index";

const mockProps: HeaderProps = {
  selectedCategory: StoryCategories.Health,
  handleClick: jest.fn(),
};

describe("<BottomSection />", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should match the BottomSection component snapshot", () => {
    const view = render(<BottomSection {...mockProps} />);
    expect(view.container).toMatchSnapshot();
  });
});
