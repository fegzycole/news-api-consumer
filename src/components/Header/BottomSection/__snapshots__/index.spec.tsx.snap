// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BottomSection /> should match the BottomSection component snapshot 1`] = `
<div>
  <nav
    class="Categories_Container"
  >
    <ul
      class="Categories"
    >
      <li
        class="Category"
      >
        <button
          class="Category_Btn false"
        >
          Top Headlines
        </button>
      </li>
      <li
        class="Category"
      >
        <button
          class="Category_Btn false"
        >
          Business
        </button>
      </li>
      <li
        class="Category"
      >
        <button
          class="Category_Btn false"
        >
          Entertainment
        </button>
      </li>
      <li
        class="Category"
      >
        <button
          class="Category_Btn false"
        >
          General
        </button>
      </li>
      <li
        class="Category"
      >
        <button
          class="Category_Btn Selected"
        >
          Health
        </button>
      </li>
      <li
        class="Category"
      >
        <button
          class="Category_Btn false"
        >
          Science
        </button>
      </li>
      <li
        class="Category"
      >
        <button
          class="Category_Btn false"
        >
          Sports
        </button>
      </li>
      <li
        class="Category"
      >
        <button
          class="Category_Btn false"
        >
          Technology
        </button>
      </li>
    </ul>
  </nav>
</div>
`;
